overlord/devicestate/devicestate.go

```diff
@@ -475,45 +476,73 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 }
 
 type remodelVariant struct {
-	localSnapsRequired bool
+	offline        bool
+	localSnaps     []*snap.SideInfo
+	localSnapPaths []string
 }
 
 type pathSideInfo struct {
 	localSi *snap.SideInfo
 	path    string
 }
 
-func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State,
-	pathSI *pathSideInfo, snapName string, opts *snapstate.RevisionOptions,
+func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State, snapName string, snapID string, opts *snapstate.RevisionOptions,
 	userID int, snapStateFlags snapstate.Flags, tracker snapstate.PrereqTracker,
 	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
 	logger.Debugf("snap %s track changed", snapName)
 	if opts == nil {
 		opts = &snapstate.RevisionOptions{}
 	}
-	if ro.localSnapsRequired {
-		if pathSI == nil {
-			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
+	if ro.offline {
+		pathSI := sideInfoAndPathFromID(ro.localSnaps, ro.localSnapPaths, snapID)
+		if pathSI != nil {
+			return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
+				userID, snapStateFlags, tracker, deviceCtx, fromChange)
 		}
-		return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
-			userID, snapStateFlags, tracker, deviceCtx, fromChange)
+
+		info, err := snapstate.CurrentInfo(st, snapName)
+		if err != nil {
+			if errors.Is(err, &snap.NotInstalledError{}) {
+				return nil, fmt.Errorf("no snap file provided for %q", snapName)
+			}
+			return nil, err
+		}
+
+		if opts != nil && !opts.Revision.Unset() && info.Revision != opts.Revision {
+			return nil, fmt.Errorf("installed snap %q does not match revision required to be used for offline remodel: %s != %s", snapName, opts.Revision, info.Revision)
+		}
+
+		// this would only occur from programmer error, since
+		// UpdateWithDeviceContext should only be called if either the snap
+		// revision or channel needs to change
```

Comment:
but this cannot be a revision change as the revision is current one?

Reply:
Once we get to this point, yes. But the comment was referring to the cases in which `UpdateWithDeviceContext` would be called, which are revision and channel changes.